### DISCUSSION TITLE
Allow interconnecting multiple lab instances

### DISCUSSION THREAD

**Discussion by sk2** [srl-labs/containerlab#1440] [Ideas]
Hi,

We are using containerlab to model large topologies with machines that can be slow to boot and/or have convergence times.
We'd like to be able to boot multiple labs and have them interconnect. For instance being able to bring up a core lab, and then attach/detach edge or DC labs to this.

Some sort of parameterised/variable to specify the interconnection points in the various .yaml files would work.

This would be very helpful for large models.

Thanks
Simon

**Comment by steiler** [comment]
Hi @sk2,
I think what you can do is use the Kind called ext-container.
This is simply a reference in the topology to a container that was spun-up externally, meaning not by the active running clab instance.
Once references in under nodes, you can simply use that ext-container node in the Links section to provision link.
surely you can also deploy a topology that contains only ext-container nodes. in which case clab will provide its wiring capabilities only.

https://containerlab.dev/manual/kinds/ext-container/
as you can see, you can also execute commands in these containers via the exec parameter. 

Hope that helps.

**Comment by sk2** [comment]
Hi, thanks! Is it possible to use multiple of these external containers to write between devices in different topologies?
If so that would work perfectly. Thanks!

**Reply by steiler** [reply]
To your actual topology everything is external thats not defined in there. It will query the runtime e.g. docker and try to find the container with the given name. if it is there it will do the wiring ... if not .... it will wait and poll the runtime until the container is present.
There is a timeout of a couple of minutes within which the ext-container should show up. otherwise the deploy will fail.
So yes interconnecting any topology, and also non containerlab deployed containers is perfectly possible.

**Reply by sk2** [reply]
Thanks for the quick reply. 
I think our use case is slightly different, and maybe has some challenges with the timeout. 

Essentially it would be to create half of the wire, and just have that there. Then the other topology contains the other half of the wire, and then when both are booted then they are connected. 

Ideally these could be labelled to allow them to interconnect. 

It might be easier if I draw up a diagram.

**Reply by steiler** [reply]
Thats not how it works.
We do create a veth pair for a standard link definition. The veth pairs are created in the host system (root network namespace) with a random "clab-xyz" name. Then they are pushed into the namespaces you try to connect (which would e.g. be container <-> container or container <-> host, whilst in the latter case one side of the veth pair remains in the root namespace). After being pushed into the desired namespaces they are renamed to their final given name via the topology file.

Hence you cannot create just one half of the veth Pair ... it is both or nothing.
My suggestion is to deploy the main part of your topo, then deploy the second topology. And finally in the third topology you reference the containers that need interconnection as ext-containers and define the wiring.

**Reply by steiler** [reply]
It is even 15 min we're waiting... 
https://github.com/srl-labs/containerlab/blob/main/runtime/runtime.go#L109-L110

Anyways, using a shell script that spins up the things one after the other would result in basically no wait time at all for the interconnect part.

**Reply by hellt** [reply]
@sk2 the ext-container might work if your NOS accepts hot-plugging of interfaces. If it doesn't then you better use bridges to stitch labs together.

e.g. your core lab is deployed with links connected to `brIX`. Then DC lab boots up and also connects to `brIX`.
That way you have disjoint lab boot management, yet they able to talk to one another via bridges that stitch on L2 their interfaces

**Reply by sk2** [reply]
Hi @hellt, yeah I think the bridge approach is closer to what I had in mind. It's something along the lines of being able to create the bridge segment and then later refer to that in another lab and somehow magically they get stitched together.
Some way we could use an alias to the bridge on both labs, and then container lab resolves this alias to a concrete bridge id internally.

I might be mixing up the limitations of the bridge approach vs what I did in a previous system.... can the bridge be instantiated with just one end? Or does it need both the A and B ends when it is created? In a previous product we used the udp socket from libvirt so we could rewire on-the-fly using a udp mux - maybe that's not possible using bridges?

**Reply by hellt** [reply]
@sk2 you definitely can connect as many interfaces to a bridge, starting with none or just one. 
Say you have a bridge br0. You can connect node1:eth1 to br0 and later connect other interfaces (including vxlan) to it
